addons/account/models/account_invoiceable_order.py

```diff
@@ -0,0 +1,192 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import Command, _, api, fields, models
+from odoo.tools import float_is_zero
+
+
+class InvoiceableOrderMixin(models.AbstractModel):
+    _name = 'account.invoiceable.order'
+    _description = 'Enables downpayment support for models that can have down payments on them'
+
+    amount_to_invoice = fields.Monetary("Amount to invoice", store=True, compute='_compute_amount_to_invoice')
+    amount_invoiced = fields.Monetary("Already invoiced", compute='_compute_amount_invoiced')
+
+    # To override
+    company_id = fields.Many2one('res.company')
+    currency_id = fields.Many2one('res.currency')
+    fiscal_position_id = fields.Many2one('account.fiscal.position', string='Fiscal Position')
+    partner_id = fields.Many2one('res.partner')
+
+    amount_total = fields.Monetary()
+
+    invoice_ids = fields.Many2many('account.move')
+    order_line = fields.One2many('account.invoiceable.line')
+
+    invoice_status = fields.Selection([])
+
+    def _qty_field_name(self):
+        raise NotImplementedError # To override
+
+    def _sign(self):
+        raise NotImplementedError # To override
+
+    def _is_locked(self):
+        raise NotImplementedError # To override
+
+    def _filter_subrecords(self):
+        raise NotImplementedError # To override
+
+    def _subrecord_type(self):
+        raise NotImplementedError # To override
+
+    def _prepare_invoice(self):
+        raise NotImplementedError # To override
+
+    def _create_invoices(self, grouped=False, final=False, date=None):
+        raise NotImplementedError # To override
+
+    def action_view_invoice(self, invoices=False):
```

Comment:
action_view_invoice has to choose a different tree view to open between sale and purchase, so it makes sense to have a hook to override for this. Though, it also sets default values in the context in sale than in purchase ; shouldn't these behaviors be aligned (putting nearly everything into the mixin) ? I don't see why we need those default values in sale tbh :thinking: